### PR TITLE
Check override conflicts at module install

### LINKED ISSUE #36617: Module with conflicting override can still be installed
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When installing a module, if an error occurs during the process, the module is still uploaded into the shop and saved into the db. However, it is not installed. When you retry the installation of the same module, since it's present in the modules/ the install succeed without causing any errors.

### Expected behavior

If an error occurs during a module install, we shouldn't be able to install the exact same module.
Also if an error occurs during the process, the module shouldn't not be in the shop.
[screencast.webm](https://github.com/user-attachments/assets/d64983c6-7e70-47fe-b0d1-b17c9685315e)

### Steps to reproduce

1. Create or get two diffrent modules with the same override doing different actions,
2. Go to module manager and upload **one**,
3. After completion, install the **second** module,
4. An error should appear in the top right corner of your window (toaster notification),
5. You can check in db, module table, the second module should be there,
6. Try reuploading the **second** module,
7. The install should be successful.

The problems comes from the ```install()``` function in ```classes/module/Module.php```
The module is inserted into db but the override check is done after that.

### PrestaShop version(s) where the bug happened

8.1.x / 9

### PHP version(s) where the bug happened

8.1

### If your bug is related to a module, specify its name and its version

_No response_

### Your company or customer's name goes here (if applicable).

Prestashop SA

### PR DESCRIPTION
**Pull request by ludoviccardinale**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch           | 8.2.x
| Description      | This Pull Request introduces new functions to the PrestaShop core that aim to resolve a critical issue related to module overrides. The problem occurs when a module is installed, and it attempts to override a class or method that is already overridden by another module. This can lead to conflicts, unexpected behavior, and errors during the installation process.
| Type             | bug fix
| Category         | BO
| BC breaks        | no
| Deprecations     | no
| How to test      | see below
| UI Tests          | https://github.com/florine2623/testing_pr/actions/runs/10918924024
| Fixed issue or discussion     | https://github.com/PrestaShop/PrestaShop/issues/36617
| Related PRs       | 
| Sponsor company   | Prestashop SA

# HOW TO TEST

- Install module pscsx3241, installation should be succesful
- Install module pscsx32412, installation should fail
- check in db, in table `ps_module` that you have  pscsx3241
- check in db, in table `ps_module` that you *don't* have module pscsx32412

 
[pscsx3241.zip](https://github.com/user-attachments/files/17025420/pscsx3241.zip)
[pscsx32412.zip](https://github.com/user-attachments/files/17025421/pscsx32412.zip)

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hello @ludoviccardinale!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Modules.Notification`
    - [ ] `The override file %1$s conflicts with an existing override in %2$s.`

(Note: this is an automated message, but answering it will reach a real human)

**Review by Hlavtox — COMMENTED** [review]
Does this mean that two modules with the same override cannot be installed? That is a BC break, because if I am not wrong, a professional user or a module support was able to install this, after manually merging the override.

Ping @kpodemski @ShaiMagal

**Comment by kpodemski** [comment]
@Hlavtox if it compare override functions, and if there is one that already has been overridden then it fails, I think it makes sense to block the installation, you can then ask support to merge overrides and install the module without side effects

**Review by FabienPapet — CHANGES_REQUESTED** [review]
Missing types and return types. 

Perhaps some code like getClassMethodsFromContent should also be moved into dedicated files so it can be unit-tested

File: classes/module/Module.php [hunk 1]
```diff
@@ -472,6 +477,56 @@ public function install()
...
+                        'Admin.Modules.Notification'
+                    );
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+    
+    protected function compareOverrideFiles($module_override_path, $prestashop_override_path)
```
Comment (FabienPapet) [inline classes/module/Module.php:504]: should be private by default, as it is only used by our code. and does not need to be overriden

File: classes/module/Module.php [hunk 2]
```diff
@@ -472,6 +477,56 @@ public function install()
...
+    
+        foreach ($module_methods as $method) {
+            if (in_array($method, $prestashop_methods)) {
+                return false;
+            }
+        }
+        return true;
+    }
+    
+    protected function getClassMethodsFromContent($content)
```
Comment (FabienPapet) [inline classes/module/Module.php:517]: should be private by default, as it is only used by our code. and does not need to be overriden

File: classes/module/Module.php [hunk 3]
```diff
@@ -472,6 +477,56 @@ public function install()
         return true;
     }
 
+    protected function checkOverrideConflicts()
```
Comment (FabienPapet) [inline classes/module/Module.php]:
```suggestion
    protected function checkOverrideConflicts(): bool
```

File: classes/module/Module.php (hunk 1 above)
Comment (FabienPapet) [inline classes/module/Module.php]:
```suggestion
    private function compareOverrideFiles(string $module_override_path, string $prestashop_override_path)
```

File: classes/module/Module.php (hunk 2 above)
Comment (FabienPapet) [inline classes/module/Module.php]:
```suggestion
    private function getClassMethodsFromContent(string $content): array
```

**Review by FabienPapet — DISMISSED** [review]

**Review by ShaiMagal — CHANGES_REQUESTED** [review]
In every case this should make problems, it can't be in branch 8.x, must be for develop branch. At least, but better not approve this... I don't like this "solution" :(

If I understand this well.... -> 
Now situation (before apply PR): Module can't be enabled and you see error message, and eshop owner know, where the problem is. And eshop owner can contact developer.

Situation after apply this PR: Module can't be installed and you **don't see reason**, only fail... So.... **No one know there is problem with override...** - it will waste developer time for investigating.

**This PR will make more evil than good. If this is needed, there must be any message what tell eshop owner, why module can't be installed.....**

Now situation is OK for me, if eshop owner can't install (enable) second module with same override, he will be notified about override problem and he can contact module developer to manual fix.

I prefer to not approve this PR, or at least there must be any modification with error message "why module can't be installed".

**Comment by kpodemski** [comment]
@ShaiMagal 

It seems that you didn't understand this change. 

> Now situation (before apply PR): Module can't be enabled and you see error message, and eshop owner know, where the problem is. And eshop owner can contact developer.

No, the current situation is that even if there are errors with overrides, the module is silently installed. With reinstallation of the module, the original override (from another module) may be removed mistakenly.

> Situation after apply this PR: Module can't be installed and you don't see reason, only fail... So.... No one know there is problem with override... - it will waste developer time for investigating.

No. You see the reason, there will be a message about conflicting override:

```
$this->_errors[] = Context::getContext()->getTranslator()->trans(
                        'The override file %1$s conflicts with an existing override in %2$s.',
                        [$module_override_path, $prestashop_override_path],
                        'Admin.Modules.Notification'
                    );
```

@ludoviccardinale PHPCsfixer is unhappy with some of the styling of the code. Could you fix it?

**Comment by kpodemski** [comment]
@Hlavtox 

It's not a bc break. With the situation you described:

> Does this mean that two modules with the same override cannot be installed? That is a BC break, because if I am not wrong, a professional user or a module support was able to install this, after manually merging the override.

You have to disable overrides from one module anyway, meaning:
- you try to install a module and it fails
- the module is silently installed
- you contact support because it does not work
- you merge overrides
- you change `/modules/conflicting_module/override` to `/modules/conflicting_module/override-off`
- you install it again

Now, the process would be:
- you try to install a module and it fails with an error 
- you contact the support
- and all the steps above

So, eventually, it will help with a module being silently installed even if it should not be installed.

**Comment by matthieu-rolland** [comment]
I'll be in charge of this PR from now on @kpodemski 

> Missing types and return types.
> 
> Perhaps some code like getClassMethodsFromContent should also be moved into dedicated files so it can be unit-tested

yes indeed, I'll do it 👍

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop` or `8.1.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
### About linked issues
Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

**Comment by matks** [comment]
Hi, I have changed the target branch to 8.2.x because of https://github.com/PrestaShop/PrestaShop/issues/36758

**Comment by ShaiMagal** [comment]
@Hlavtox WDYT?

**Review by kpodemski — DISMISSED** [review]

**Comment by ShaiMagal** [comment]
@kpodemski I tested this PR again, and i still don't see any benefit from this.

**State:**
I installed module "A" with override Product.php.
And trying to work with module "B" with same override.

**without PR:**
A) I can't install module "B" with same override
B) If I uninstall module "A", override is deleted

**with PR:**
A) I can't install module "B" with same override
B) If I uninstall module "A", override is deleted

Same behaviour. Only other text message...

What is difference from actual state? What is benefit? What am I missing?

Then I will remove my lock of course. I like moving PrestaShop forward, but I really don't see any benefit. Please, someone explain for me.

**Comment by kpodemski** [comment]
@ShaiMagal 

Have you checked the code? For this scenario:

> without PR:
> A) I can't install module "B" with same override
> B) If I uninstall module "A", override is deleted

after installing module B with the same override, check the database and see that the module has been added to the `PREFIX_module` table

with the PR, that shouldn't be the case, and this difference is important

installing overrides happens when you here:

```
if (!$this->enable()) {
            return false;
        }
```

so, as you can see in the code, this won't be reached if there is a problem with overrides

**Comment by matthieu-rolland** [comment]
(pr not ready)

File: classes/module/Module.php [hunk 4]
```diff
@@ -472,6 +477,60 @@ public function install()
...
 
+    protected function checkOverrideConflicts(): bool
+    {
+        if (!is_dir($this->getLocalPath() . 'override')) {
+            return true;
+        }
+
+        foreach (Tools::scandir($this->getLocalPath() . 'override', 'php', '', true) as $file) {
+            $module_override_path = $this->getLocalPath() . 'override/' . $file;
+            $prestashop_override_path = _PS_ROOT_DIR_ . '/override/' . $file;
```
Comment (ShaiMagal) [inline classes/module/Module.php]: There should be `_PS_OVERRIDE_DIR_` used, not _PS_ROOT_DIR_ . '/override/'

**Comment by ShaiMagal** [comment]
@kpodemski I still don't understand, what is benefit of this change...

Are we sure, this change is really suited for PS 8.2? Not develop branch?

Are we 100% sure, there will be no impact? I prefer to add this to develop, what do you thing? For develop I don't have problem with this change.

Anyway, function getClassMethodsFromContent is strange for readability and understanding, at least, for me. It must be detaily tested, if this i for PS 8.2...  I prefer develop branch.

**This behaviour is same for years, I don't think, it's necessary to "fix" it for PS 8.2, there can be any negative impact of this change.**


Can we change branch to develop?

What do you think? :)

Then we can approve it without problems. But we need do small changes in code at least. I change my review.

File: classes/module/Module.php [hunk 5]
```diff
@@ -472,6 +477,60 @@ public function install()
...
+        foreach ($module_methods as $method) {
+            if (in_array($method, $prestashop_methods)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private function getClassMethodsFromContent(string $content): array
```
Comment (ShaiMagal) [inline classes/module/Module.php]: There must be any explanation how function works. Not easily readable for everyone :)

**Review by ShaiMagal — DISMISSED** [review]
I prefer to change branch to "develop".

Then I can approve it, but if other colleges will be agains it, of course I will not blocking PR if other commitors will have other opinion than me :)

**Comment by kpodemski** [comment]
@ShaiMagal I think I already explained the change in behavior and why it is a right thing to improve installation process. Installation fails and you still end up having a module... installed :)

Let's wait for @matthieu-rolland to continue with this PR.

**Comment by matthieu-rolland** [comment]
I just pushed a refacto, and added comments on `getClassMethodsFromContent`, whose code is not obvious.

I also had to modify `getClassMethodsFromContent` because I had edge cases were conflicts were not detected (when there's spaces before function name for example)

I'll add tests soon and fix integration tests, still wip

File: classes/module/Module.php (hunk 5 above)
Comment (matthieu-rolland) [reply classes/module/Module.php]: done

File: classes/module/Module.php (hunk 4 above)
Comment (matthieu-rolland) [reply classes/module/Module.php]: done

**Comment by matthieu-rolland** [comment]
Ready for review

File: tests/Resources/modules_tests/testbasicconflict/testbasicconflict.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public function install()
+    {
+        if (parent::install() == false) {
+            return false;
+        }
+
+        return true;
+    }
```
Comment (kpodemski) [inline tests/Resources/modules_tests/testbasicconflict/testbasicconflict.php]:
```suggestion
```

not needed

File: tests/Resources/modules_tests/testtrickyconflict/testtrickyconflict.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public function install()
+    {
+        if (parent::install() == false) {
+            return false;
+        }
+
+        return true;
+    }
```
Comment (kpodemski) [inline tests/Resources/modules_tests/testtrickyconflict/testtrickyconflict.php]:
```suggestion
```

not needed

File: src/Core/Module/ModuleOverrideVerifier.php [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+            if (file_exists($prestashop_override_path)) {
+                if ($this->hasConflictingMethod($module_override_path, $prestashop_override_path)) {
+                    $this->errors[] = $this->translator->trans(
+                        'The override file %1$s conflicts with an existing override in %2$s.',
+                        [$module_override_path, $prestashop_override_path],
+                        'Admin.Modules.Notification'
+                    );
+                }
+            }
+        }
```
Comment (kpodemski) [inline src/Core/Module/ModuleOverrideVerifier.php]:
```suggestion
        foreach ($tools->scandir($moduleOverridePath, 'php', '', true) as $file) {
            $moduleOverrideFile = $moduleOverridePath . DIRECTORY_SEPARATOR . $file;
            $existingOverrideFile = _PS_OVERRIDE_DIR_ . $file;

            if (file_exists($existingOverrideFile)) {
                if ($this->hasConflictingMethod($moduleOverrideFile, $existingOverrideFile)) {
                    $this->errors[] = $this->translator->trans(
                        'The override file %1$s conflicts with an existing override in %2$s.',
                        [$moduleOverrideFile, $existingOverrideFile],
                        'Admin.Modules.Notification'
                    );
                }
            }
        }
```

File: src/Core/Module/ModuleOverrideVerifier.php [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        $prestashop_methods = $this->getClassMethodsFromContent(file_get_contents($prestashop_override_path));
+
+        foreach ($module_methods as $method) {
+            if (in_array($method, $prestashop_methods)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
```
Comment (kpodemski) [inline src/Core/Module/ModuleOverrideVerifier.php]:
```suggestion
    private function hasConflictingMethod(string $moduleOverridePath, string $existingOverridePath): bool
    {
        $moduleMethods = $this->getClassMethodsFromContent(file_get_contents($moduleOverridePath));
        $existingOverrideMethods = $this->getClassMethodsFromContent(file_get_contents($existingOverridePath));

        foreach ($moduleMethods as $method) {
            if (in_array($method, $existingOverrideMethods)) {
                return true;
            }
        }

        return false;
    }
```

File: src/Core/Module/ModuleOverrideVerifier.php [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+                // collect method name
+                if ($method_name) {
+                    $methods[] = $method_name; // Save method name
+                }
+            }
+        }
+
+        return $methods;
+    }
```
Comment (kpodemski) [inline src/Core/Module/ModuleOverrideVerifier.php]:
```suggestion
    private function getClassMethodsFromContent(string $content): array
    {
        // Get the list of tokens, see https://www.php.net/manual/en/function.token-get-all.php
        $tokens = token_get_all($content);
        $methods = [];
        foreach ($tokens as $index => $token) {
            $methodName = null;
            // filter method definitions
            if (is_array($token) && $token[0] === T_FUNCTION) {
                // This loop is necessary in case some additional spaces exist before the function name
                for ($i = $index + 1; $i < count($tokens); ++$i) {
                    if (is_array($tokens[$i]) && $tokens[$i][0] === T_STRING) {
                        $methodName = $tokens[$i][1];
                        break;
                    }
                }

                // Collect method name
                if ($methodName) {
                    $methods[] = $methodName; // Save method name
                }
            }
        }

        return $methods;
    }
```

**Review by kpodemski — COMMENTED** [review]
Non-blocking suggestions

File: tests/Resources/modules_tests/testbasicconflict/testbasicconflict.php (hunk 6 above)
Comment (matthieu-rolland) [reply tests/Resources/modules_tests/testbasicconflict/testbasicconflict.php]: ah yes, I duplicated other test modules to use them as templates, didn't catch this 👍

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Modules.Notification`
    - [ ] `The override file %1$s conflicts with an existing override in %2$s.`

(Note: this is an automated message, but answering it will reach a real human)

**Review by kpodemski — DISMISSED** [review]

File: composer.json [hunk 11]
```diff
@@ -244,7 +244,10 @@
         ],
         "unit-tests": [
             "@php -d date.timezone=UTC ./vendor/phpunit/phpunit/phpunit -c tests/Unit/phpunit.xml"
-        ]
+        ],
+        "php-cs-fixer": "@php ./vendor/bin/php-cs-fixer fix --path-mode=intersection",
```
Comment (matthieu-rolland) [inline composer.json]: Added php-cs-fixer (and phpstan as I'm at it) commands in the composer.json, so that we can easily launch it with the same parameters used by the CI

**Review by kpodemski — DISMISSED** [review]

File: classes/module/Module.php [hunk 12]
```diff
@@ -411,6 +412,17 @@ public function install()
             return false;
         }
 
+        // Check for override conflicts
+        $moduleOverrideVerifier = $this->get(ModuleOverrideVerifier::class);
```
Comment (jolelievre) [inline classes/module/Module.php]:
```suggestion
        $moduleOverrideChecker = $this->get(ModuleOverrideChecker::class);
```
`Verifier` sounds like an anglicism no?

Comment (jolelievre) [reply classes/module/Module.php]: Also should this be inculded in the `enable` method? It depends if the override classes are generated only on install or also on enable? (it would make sense that they are done on both actions, it also means on disable the overrides should be cleaned)

File: composer.json [hunk 13]
```diff
@@ -244,14 +244,20 @@
...
+        "php-cs-fixer:dry": "@php ./vendor/bin/php-cs-fixer fix --dry-run --diff --path-mode=intersection",
+        "phpstan": "@php ./vendor/bin/phpstan analyze -c phpstan.neon.dist"
```
Comment (jolelievre) [inline composer.json]: One last interesting command would be `phpstan-baseline` that updates the baseline for exceptions with the `--generate-baseline` parameter

File: src/Core/Module/ModuleOverrideVerifier.php [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Module;
+
+use PrestaShop\PrestaShop\Adapter\Tools;
```
Comment (jolelievre) [inline src/Core/Module/ModuleOverrideVerifier.php]: We should try not to use adapter classes in core code, this scandir method could be added in a core util class, or maybe supply use the `Finder` symfony component to do the same thing

File: src/Core/Module/ModuleOverrideVerifier.php [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+        // if module doesn't have an override folder, there is no conflict
+        if (!is_dir($moduleOverridePath)) {
+            return false;
+        }
+
+        $tools = new Tools();
+        // module has overrides, let's check override files one by one
+        foreach ($tools->scandir($moduleOverridePath, 'php', '', true) as $file) {
+            $moduleOverrideFile = $moduleOverridePath . DIRECTORY_SEPARATOR . $file;
+            $existingOverrideFile = _PS_OVERRIDE_DIR_ . $file;
```
Comment (jolelievre) [inline src/Core/Module/ModuleOverrideVerifier.php]: The override directory should be injected in the constructor, this will make the code more adaptable and easier to unit test

File: classes/module/Module.php (hunk 12 above)
Comment (matthieu-rolland) [reply classes/module/Module.php]: ah damn yes, should have been `verifyer` 🙈

ok for `checker`

I'm not sure about how overrides are dealt with when enabling the module, I'll look into it

File: src/Core/Module/ModuleOverrideVerifier.php (hunk 15 above)
Comment (matthieu-rolland) [reply src/Core/Module/ModuleOverrideVerifier.php]: done 👍

File: classes/module/Module.php (hunk 12 above)
Comment (matthieu-rolland) [reply classes/module/Module.php]: done 👍

File: src/Core/Module/ModuleOverrideVerifier.php (hunk 14 above)
Comment (matthieu-rolland) [reply src/Core/Module/ModuleOverrideVerifier.php]: I used Symfony finder instead 👍

**Review by jolelievre — DISMISSED** [review]
Thanks @ludoviccardinale and @matthieu-rolland

**Review by jolelievre — DISMISSED** [review]

**Review by jolelievre — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Review by boherm — APPROVED** [review]

**Review by florine2623 — APPROVED** [review]
Hello @ludoviccardinale and @matthieu-rolland ,

I followed the How to test. Here's the result :

**Before :** 

Install module pscsx3241, installation should be successful ✅ 
Install module pscsx32412, installation should fail ✅ 
check in db, in table ps_module that you have pscsx3241 ✅ 
check in db, in table ps_module that you don't have module pscsx32412 :x:

https://github.com/user-attachments/assets/3657a394-8d90-4bc7-b8de-d840b61d6677

Screenshot 2024-09-18 at 12 20 47
Screenshot 2024-09-18 at 12 25 22


**After :**

Install module pscsx3241, installation should be successful ✅ 
Install module pscsx32412, installation should fail ✅ 
check in db, in table ps_module that you have pscsx3241 ✅ 
check in db, in table ps_module that you don't have module pscsx32412 ✅ 

https://github.com/user-attachments/assets/b8b8830e-a7e7-4ebe-a68b-1de18f6a00bb

Screenshot 2024-09-18 at 12 30 49

Module is not installed so not in database :
Screenshot 2024-09-18 at 12 31 26


Also tested the deletion of both modules, checked the database, no error/exceptions were raised on BO, FO is accessible. 


It is a GO for QA ^^
Thanks!
